### PR TITLE
An attempt to create an "Adapton Functor", as a class, in Motoko

### PR DESCRIPTION
**Pull request by matthewhammer**
This PR is substantial.

_Before:_ CleanSheets defined adapton and the DSL for formula as one set of mutually-recursive definitions.

_Now:_ CleanSheets and its DSL are entirely abstract, and can be replaced with any other DSL implementation, which it will help incrementalize (e.g., [IC-LOGO](https://github.com/chenyan2002/ic-logo)).

### CODE REVIEW

File: src/types.mo [hunk 1]
```diff
@@ -7,213 +7,29 @@ import L "mo:stdlib/list";
 
 module {
 
-public module Sheet {
+// A generic Adapton engine is parameterized by these choices:
+public type EvalSig = module {
+  type Name;
+  nameEq : (n1:Name, n2:Name) -> Bool;
```
Comment (matthewhammer) [inline src/types.mo]: @crusso I recall you saying that we cannot have type members in the module types, forbidding what I'm attempting to say above, right?

Comment (crusso) [reply src/types.mo]: Yup.

Comment (crusso) [reply src/types.mo]: It's just an unimplemented feature, not a fundamental restriction I hope.
For now, you could either parameterize EvalSig on the type componenents, assuming they aren't higher-kinded, or you could try to define EvalSig indirectly, using a module class, which should export the type components from the body. 
I'll try that on a small example to verify.

Comment (matthewhammer) [reply src/types.mo]: Ah, the module class idea hadnt occured to me.  I don't see it yet, but I also haven't yet tried either.

File: src/adapton2.mo [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import P "mo:base/prelude";
+
+import T "types";
+
+module {
+
+public class AdaptonEngine<_Name, _Val, _Error, _Env, _Exp>(Eval:T.Eval<_Name, _Val, _Error, _Env, _Exp>) {
+
+  // Compiler says: type error, type definition Val = _Val references type parameter(s) _Val from an outer scope
+  type Val = _Val;
```
Comment (matthewhammer) [inline src/adapton2.mo]: @crusso I am not able to define any internal types using the type parameters here.

How did we not run into this issue this morning?  We didn't try this pattern?

Comment (crusso) [reply src/adapton2.mo]: Yeah we didn't do the re-export. I had 2 PRs that fix this but Andreas didn't like either. For now. Just don't declare those type components and it should work

File: src/adapton2.mo [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+import R "mo:base/result";
+import P "mo:base/prelude";
+
+import E "evalType";
+
+module {
+
+  // morally, there are 5 user-defined types over which this module is
+  // parameterized: Env, Exp, Val, Error and Name.
+  // Some types defined below require all five of these abstract parameters.  Some require fewer.
```
Comment (matthewhammer) [inline src/adapton2.mo]: @rossberg This is where @crusso was helping me navigate the design space, given the limitations of how the scope of type variables' is currently restricted.

Ideally, many of these types defined just below would be able to move within the class defined below, and be private to it.

File: src/adapton2.mo [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+import H "mo:base/hashMap";
+import Hash "mo:base/hash";
+import Buf "mo:base/buf";
+import L "mo:base/list";
+import R "mo:base/result";
+import P "mo:base/prelude";
+
+import E "evalType";
+
+module {
```
Comment (matthewhammer) [inline src/adapton2.mo]: @crusso Here's what happened when I did what we agreed to try.

File: src/adapton2.mo [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+      case (_, _) {
+             false
+           }
+      }
+
+    };
+
+
+    public func init(_logFlag:Bool) : Context {
+      // to do -- compiler bug? -- IR typing issue when this line is inlined to its use below:
```
Comment (crusso) [inline src/adapton2.mo]: is that bug still there?

Comment (matthewhammer) [reply src/adapton2.mo]: Nope, this must have been from an old compiler run, by mistake.

**Pull request merged**